nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp

```diff
@@ -232,24 +197,128 @@ class DynamicParamsClient
   }
 
   // A check to filter whether parameter name is part of the lastest event
-  bool is_in_event(const std::string & name)
+  bool is_in_event(const std::string & path, const std::string & param_name)
   {
-    rclcpp::ParameterEventsFilter filter(last_event_, {name},
+    auto full_path = path;
+    if (*full_path.begin() != '/') {
+      full_path = '/' + full_path;
+    }
+
+    rclcpp::ParameterEventsFilter filter(last_event_, {param_name},
       {rclcpp::ParameterEventsFilter::EventType::NEW,
         rclcpp::ParameterEventsFilter::EventType::CHANGED});
-    return !filter.get_events().empty();
+
+    return full_path == last_event_->node && !filter.get_events().empty();
   }
-  // Variant of is_in_event to also check the node path of last event
-  bool is_in_event(const std::string & path, const std::string & name)
+
+  // Variant of is_in_event to specify namespace and node name
+  bool is_in_event(
+    const std::string & name_space,
+    const std::string & node_name, const std::string & param_name)
   {
-    auto full_path = path;
-    if (*full_path.begin() != '/') {
-      full_path = '/' + full_path;
+    auto full_path = join_path(name_space, node_name);
+    return is_in_event(full_path, param_name);
+  }
+
+  // Variant of is_in_event to check under member node
+  bool is_in_event(const std::string & param_name)
+  {
+    return is_in_event(node_->get_namespace(), node_->get_name(), param_name);
+  }
+
+protected:
+  void event_callback(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+  {
+    last_event_ = event;
+    if (is_event_in_map(event)) {
+      user_callback_();
     }
-    return full_path == last_event_->node && is_in_event(name);
   }
 
 private:
+  // Get current parameters from remote or member node
+  // For remote nodes, will re-try to grab parameters up to a maximum # of attempts
+  std::vector<rclcpp::Parameter> get_params(
+    const std::string & path,
+    const std::vector<std::string> & param_names, int attempts_max = 5)
+  {
+    std::vector<rclcpp::Parameter> params;
+    bool success = false;
+    int attempts = 0;
+    if (!params_from_this(path, param_names, params)) {
+      while (!success && attempts < attempts_max) {
+        success = params_from_remote(path, param_names, params);
+        attempts++;
+      }
+    }
+    return params;
+  }
+
+  // Get current parameters from member node
+  bool params_from_this(
+    const std::string & path,
+    const std::vector<std::string> & param_names, std::vector<rclcpp::Parameter> & params)
+  {
+    if (path == join_path(node_->get_namespace(), node_->get_name())) {
```

Comment:
I think I'd pull the join_path into a separate function - getNodePath(node_). Then I'd pull this line out of this function  and move it to line 248, so it reads more like
```
if path == getNodePath(node_) // does path refer to this node
  params_from_this()
else
  while ....
```

That way the params from remote and params from this have the same structure. As it is right now, they both return a bool but the meaning of that bool is quite different. In the first the bool means the path doesn't refer to this node. In the other, it means communication was succesful or failed.